### PR TITLE
Feature/experiment

### PR DESCRIPTION
**Pull request by Atahanak**
Initial PR for the Experiment namespace. The experiment infrastructure is implemented along with some additional example functions for the user and two lib examples.

Experiment tutorial and advanced testing left to another PR.

### CODE REVIEW

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]
Looks great. But I have some comments, requests, and discussions :) 
* Add some tests if possible.
* Make `LoadDataFunction` an `std::function` rather than a function pointer. This allows for us to pass formats to the loader from main as shown here (I compiled this and it worked):
    image
    Note: this would require the `_dataLoaders` member to become `vector<LoadDataFunction>`.
* Make `LoadDataFunction` take a map of strings->strings, vector of strings, or a map string->any.
    * Maybe additional strings can be part of the `std::any` parameter in the pair?
* I like how the file parameters are passed to the kernel. I think we should generalize this further:
    1. Just like how files have an associated `std::any` parameter, I think preprocessing should have one. Kernel functions would take both the file params and the preprocess params. 
    2. Pass the file params object to the preprocessing function.
    
    In other words, a file will have its params, a preprocess its params, and a kernel its params. A loader functions will know about the params of its input, a preprocess function will know about its params and the params of the file it's reading, and a kernel function will know about its params, the preprocess params used for this iteration, and the params of the input it's using. 

    Just to clarify, I mean that the same file params that you pass to the kernel function you pass to the preprocess and loaded function, and the same preprocess param you pass to the kernel you pass to the preprocessing, i.e. a single param object per function.
* I think we should have a function that generates IDs. This way, the user can generate the ID himself in case he wanted to query something. This is not that important though.

File: src/sparsebase/experiment/experiment.h [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+namespace sparsebase::experiment {
+
+//! Function template for loading data to the experiment.
+//! All data that is going to be used in the experiment must be provided via a function that follows this template.
+/*!
+ *
+ * \param file_path Path of the file to be read.
+ * \return data extracted from the file.
+ */
+using LoadDataFunction = std::unordered_map<std::string, format::Format*> (std::string & file_path);
```
[the remark this answers is not in the document]
Comment (Atahanak) [reply src/sparsebase/experiment/experiment.h:30]: I agree gonna add it in my update.

File: tests/suites/sparsebase/feature/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-add_executable(feature_tests.test feature_tests.cc)
+add_executable(feature_tests.test feature_tests.cc ../experiment/experiment_tests.cc)
```
Comment (AmroAlJundi) [inline tests/suites/sparsebase/feature/CMakeLists.txt:1]: I think this is supposed to be changed back right?

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -22,6 +21,7 @@ if(NOT ${_HEADER_ONLY} OR NOT ${USE_CUDA})
   add_subdirectory(custom_format)
   add_subdirectory(custom_converter)
   add_subdirectory(custom_experiment)
+  add_subdirectory(example_experiment)
```
Comment (AmroAlJundi) [inline examples/CMakeLists.txt:24]: I couldn't make this compile when `USE_CUDA=OFF`. Probably due to the file extension.

File: examples/example_experiment/example_experiment.cu [hunk 4]
```diff
@@ -201,9 +201,9 @@ int main(int argc, char **argv){
 
   // add kernels that will carry out the sparse matrix vector multiplication
   // init random vals large enough for all the files
-  auto vals = new float[1049866];
-  fill_r(vals, 1049866);
-  auto vals_v = Array<float>(1049866, vals);
+  auto vals = new float[2*1049866];
```
Comment (AmroAlJundi) [inline examples/example_experiment/example_experiment.cu:204]: Shouldn't this be the number of rows in dblp? i.e. 317080?

File: tests/suites/sparsebase/experiment/experiment_tests.cc [hunk 5]
```diff
@@ -22,15 +28,23 @@ auto load_data_f( vector<string> & file_names ) {
...
 TEST(experiment, Basics){
- cout << "tstss" << endl;
+  ConcreteExperiment exp;
+  exp.AddDataLoader(load_data_f, {});
+  exp.AddPreprocess("id", preprocess_f, {});
+  exp.Run(2, true);
+  auto aux = exp.GetAuxiliary();
+  for(const auto & [id, data]: aux){
+    cout << id << endl;
+
```
Comment (AmroAlJundi) [inline tests/suites/sparsebase/experiment/experiment_tests.cc:48]: The test needs some ASSERT or EXPECT statements :)

File: examples/example_experiment/example_experiment.cu (hunk 4 above)
Comment (Atahanak) [reply examples/example_experiment/example_experiment.cu:204]: nope its the number of non zeros.

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]

File: tutorials/003_experiment/start_here/tutorial_003.cc [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  default_random_engine rnd{std::random_device{}()};
+  uniform_real_distribution<float> dist(-1, 1);
+  for(unsigned int i = 0; i < size; i++){
+    arr[i] = dist(rnd);
+  }
+}
+
+int main(int argc, char **argv){
+
+  experiment::ConcreteExperiment exp;
```
[the remark this answers is not in the document]
Comment (Atahanak) [reply tutorials/003_experiment/start_here/tutorial_003.cc:86]: yes

File: tests/suites/sparsebase/experiment/experiment_tests.cc (hunk 5 above)
Comment (Atahanak) [reply tests/suites/sparsebase/experiment/experiment_tests.cc:48]: tests are incomplete for now

**Review by AmroAlJundi — APPROVED** [review]
Miniscule changes were requested, but they're not that important.

**Review by AmroAlJundi — APPROVED** [review]

**Pull request merged**